Add support for spack monitor

vsoch opened the pull request:
This provides initial support for [spack monitor](https://github.com/spack/spack-monitor), a web application that stores information and analysis about Spack installations.  Spack can now contact a monitor server and upload analysis -- even after a build is already done.

Specifically, this adds:
- [x] monitor options for `spack install`
- [x] `spack analyze` command
- [x] hook architecture for analyzers
- [x] separate build logs (in addition to the existing combined log)
- [x] docs for spack analyze
- [x] reworked developer docs, with hook docs
- [x] analyzers for:
  - [x] config args
  - [x] environment variables
  - [x] installed files
  - [x] libabigail

There is a lot more information in the docs contained in this PR, so consult those for full details on this feature.

Additional tests will be added in a future PR.

vsoch committed:
start of work to add monitor, for code and docs

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
updates to start pinging spackmon for build statuses

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding parsing of package metadata folder

this has not been tested because we need to be able to also send
packages that are for build only (currently they are not included in the
spec so the server does not know about them

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding phase level log files

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
Merge branch 'develop' of github.com:spack/spack into add/spack-monitor

vsoch committed:
updating client to send phases (needs more testing)

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
start of work to refactor using build objects in api

I still need to test the phase update, and write logic
to get build objects (and abi metadata) for server

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding tested upload for package/spec phases

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
Merge branch 'develop' of github.com:spack/spack into add/spack-monitor

vsoch committed:
formatting (and not sure about version error)

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
How do I fix this?

https://github.com/spack/spack/runs/1972561398?check_suite_focus=true

vsoch committed:
support for python2 in monitor imports

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding noqa to "double import" of urllib

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
mypy is the one that needs ignore

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
try fixing vermin issues by using sjson

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
updating spack command line completion

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
should use os.path.join

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding spack analyzers and docs

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
updating with develop

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding bootstrap example

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
Merge branch 'develop' of github.com:spack/spack into add/spack-monitor

vsoch committed:
adding working analyzer framework for libabigail to save to file

with this commit, we are adding an ability to bootstrap libabigail, wrap the abidw executable,
and then save to an xml file. Ideally we want to save to gzipped xml, but libabigail does not
work to compile this yet. The remaining work to do here is to add the monitor server
interaction. Note that we have to add python to the libabigail package.py, as without it,
nothing is created in the bin. libabigail+docs is also needed to generate anything beyond
include

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding developer docs for spack analyze, tested uploading abi xml (not compressed)

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
catching error if abidw is not able to run for libabigail analyzer

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
Current error for docs is:

```
Warning, treated as error:
/home/runner/work/spack/spack/lib/spack/docs/developer_guide.rst:458:Inline emphasis start-string without end-string.
Makefile:90: recipe for target 'html' failed
make: *** [html] Error 2
Error: Process completed with exit code 2.
```
I'm not sure how to fix it - I don't see any issue with that line, not sure what a start-string and end-string are! :thinking:

vsoch committed:
updating command completion

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
try removing *.xml (putting in code)

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
consle -> console

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
okay making progress! I'm not sure if I changed something to break this test:

![image](https://user-images.githubusercontent.com/814322/110525227-93bba300-80d1-11eb-9a1b-fd4abed2f97c.png)

vsoch committed:
exception needs to be raised on phase error

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
remove ipython

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
Merge branch 'develop' of github.com:spack/spack into add/spack-monitor

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -0,0 +1,156 @@
+    analyze                   archived-files                     repos                              spack-build-03-build-out.txt
+    spack-build-out.txt       install_environment.json           spack-build-01-autoreconf-out.txt  spack-build-04-install-out.txt
+    spack-configure-args.txt  install_manifest.json              spack-build-02-configure-out.txt   spack-build-env.txt
+    spec.yaml
```
I think this might be too wide for docs -- can you make it 80 cols?

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -0,0 +1,156 @@
+
+=======
+Analyze
+=======
```
I think it makes sense to add a top level section in `index.rst`, down below contributing and above API docs called "research", and to stick these docs in there.  Otherwise this is going to be confusing to normal users.

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -0,0 +1,156 @@
+    ==> install_files                      : install file listing read from install_manifest.json
+    ==> environment_variables              : environment variables parsed from spack-build-env.txt
+    ==> config_args                        : config args loaded from spack-configure-args.txt
+    ==> abigail                            : Application Binary Interface (ABI) features for objects
```
This UI looks a little different from the rest of Spack... doesn't really have to change right now but probably should eventually.  tty.msg is really for messages, not for stuff in lists in output.

vsoch replied:
I can change it to a print (or something else?) I just liked the blue :)

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -0,0 +1,156 @@
+
+.. code-block:: console
+
+    $ spack analyze zlib@1.2.11%gcc@7.5.0 arch=linux-ubuntu18.04-skylake
```
I'd more typically use one of these:

```
spack analyze /fz2bs56
spack analyze zlib /fz2bs56
```

As the spec you pasted here isn't guaranteed to be unique (the hash is though)

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -0,0 +1,156 @@
+=======
+
+The analyze commands allows you to take an existing package install, and run
+an analyzer to extract one or more metrics. The metrics can range from reading
```
Analyzer is *almost* defined here but not quite -- can this be changed to say that the analyze command is the front-end to various tools that let us analyze package installations?  And analyzers are modules for different types of analysis (e.g., binary, log analysis, etc.) we can do on a package installation?

One worry is that "analyzer" on its own is pretty general -- but we don't use it to mean anything else so I guess it's fine to use it for this.

Metrics always seem like scalar values to me (or several of them) -- this data seems more voluminous -- not sure what a more general term would be though.

vsoch replied:
I'll try to remove the word metric and state it another way.

tgamblin commented on lib/spack/spack/cmd/__init__.py:
```diff
@@ -112,7 +112,6 @@ def get_module(cmd_name):
     """
     require_cmd_name(cmd_name)
     pname = python_name(cmd_name)
-
```
this could be removed from the PR

tgamblin replied:
remove extra whitespace

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -0,0 +1,156 @@
+    $ ls opt/spack/linux-ubuntu20.04-skylake/gcc-9.3.0/wget-1.20.3-bum6zeaezbuqjhudzphcsyb4avkiizhj/.spack/analyze/
+    spack-analyzer-environment-variables.json  spack-analyzer-install-files.json  spack-analyzer-libabigail-wget.xml
+
+It is currently flat (not compressed) xml, and will eventually be gzipped when libabigail supports it.
```
Maybe remove the last part since they might not ever support gzipping.  Or, if we really want it gzipped, we do not have to do that via libabigail.  We can just zip it ourselves after each run -- is that worth doing?

vsoch replied:
We could, but if libabigail can't actually read the zipped content, we'd also need to unzip it as well. That would save us space in a database binary field over a full text, but require a bit of extra work (probably reasonable to do). I agree we should remove this sentence here since we don't really know, and update the docs if it eventually does output compressed.

tgamblin commented on lib/spack/docs/monitoring.rst:
```diff
@@ -0,0 +1,88 @@
+
+By default, the host for your server is expected to be at ``http://127.0.0.1``
+with a prefix of ``ms1``, and if this is the case, you can simply add the
+``--monitor`` flag to the install command:
```
The auth scheme seems simple (which is good) and I haven't thought through what we actually need here, but the defaults seem good enough for now -- we can modify them later if we need to.

tgamblin commented on lib/spack/llnl/util/tty/log.py:
```diff
@@ -440,9 +440,9 @@ def __init__(self, file_like=None, echo=False, debug=0, buffer=False,
-            buffer (bool): pass buffer=True to skip unbuffering output; note
+            echo     (bool): whether to echo output in addition to logging it
+            debug    (int): positive to enable tty debug mode during logging
+            buffer   (bool): pass buffer=True to skip unbuffering output; note
```
Can you remove the whitespace changes that are not related to this PR?

vsoch replied:
yes, sorry about that! I was originally going to update the log.py to do something with phases, but then decided that was the wrong level to do it.

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -299,6 +315,198 @@ Conceptually, packages are overloaded.  They contain:
+-----------------
+
+To write an analyzer, you can either include it in one of the files in the
+analyzers module folder (e.g., if it's associated with an application
```
where's the analyzers module folder (also should it be "directory"?)

vsoch replied:
It's at `lib/spack/spack/analyzers`, I'll add that here and change folder to directory.

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -299,6 +315,198 @@ Conceptually, packages are overloaded.  They contain:
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, when you run ``spack analyze`` an analyzer output folder will
+be created in the spec's package directory for results:
```
I think I was the one who suggested this location, but the more I think about it the more I think we want to move it.  We haven't always been good at this, but going forward we need to assume that the install tree may be read-only for pretty much any command but `install`.  e.g., in the Axom installations we've talked about analyzing, you don't actually have write access to `$prefix/.spack`, but you'll want to analyze those directories nonetheless.  We'll need a way for you to do that.

Look at `spack test` for an example of how we do it for that command -- the results go in `~/.spack`, which is really the only place we can assume is writable by the current user:

```console
(spackle):cmd> spack test results
==> No test suites with results to report
(spackle):cmd> spack test run libxml2
==> Spack test cy5f7q5u4zzm6oq3cgpyqpoj6k2b75tg
==> Testing package libxml2-2.9.10-6zobott
==> Testing package libxml2-2.9.10-dqzlvh6
(spackle):cmd> spack test results    
==> Results for test suite 'cy5f7q5u4zzm6oq3cgpyqpoj6k2b75tg':
==>   libxml2-2.9.10-6zobott PASSED
==>   libxml2-2.9.10-dqzlvh6 PASSED
(spackle):cmd> ls ~/.spack                        
cache/  config.yaml  darwin/  mirrors.yaml  test/
(spackle):cmd> ls ~/.spack/test 
compilers.yaml  cy5f7q5u4zzm6oq3cgpyqpoj6k2b75tg/
(spackle):cmd> ls ~/.spack/test/cy5f7q5u4zzm6oq3cgpyqpoj6k2b75tg 
libxml2-2.9.10-6zobott-test-out.txt*  repo/        test_suite.lock
libxml2-2.9.10-dqzlvh6-test-out.txt*  results.txt
```

I think analyzer results should probably go in there as well.  I am not sure they should be stored the same way.  Tests are stored in a way that allows you to run them many times, and the hash in the test dir is associated with the "suite" or the set of runs.

I think you want to key spack analyzer output by the same key you use for builds in the DB in spack monitor.

It would be nice to figure out if `spack test` and this can share some code, since they have very similar workflows -- tests are just another thing that can be run on already-installed packages.  Analysis isn't so different in that sense.

vsoch replied:
Hmm. For spack monitor the build_id is just an incrementing database index, and the build is unique based on the spack environment (the file that is generated in the install folder):

```
$ cat opt/spack/linux-ubuntu20.04-skylake/gcc-9.3.0/zlib-1.2.11-sl7m27mzkbejtkrajigj3a3m37ygv4u2/.spack/install_environment.json 
{
 "host_os": "ubuntu20.04",
 "platform": "linux",
 "host_target": "skylake",
 "hostname": "vanessa-ThinkPad-T490s",
 "spack_version": "0.16.1-1751-87262723ab",
 "kernel_version": "#74-Ubuntu SMP Wed Jan 27 22:54:38 UTC 2021"
}
```
and then a spec hash. So we definitely have the spec hash, but the above environment isn't really usable as a key. Are you thinking we should try to hash it or something similar? Or create a path there similar to the spec install's path?

tgamblin replied:
Yeah I think hashing this would be good.  Then you're guaranteed that whatever you add to the JSON there, you'll still be able to generate unique ids.  Although that doesn't contain the full hash.  Is it Full hash + environment?

One nitpick here: we should probably include the full commit in the spack version, for completeness.  Easiest way to do that might be to tweak `spack --version` to not truncate it.  I don't see anything terrible about doing that, though it'll make the output longer.

vsoch replied:
Hmm, so my concern here is that we want a user to be able to easily match an install to an analyzer directory, and if it's a random hash it would be meaningless. The uniqueness (at least for spack monitor) is the spec full hash plus everything in that environment, which includes the spack version.

I'll take a look at if I can figure out how to include the full commit.

vsoch replied:
I'm testing git describe and I can't figure out how to get the full commit - when I add `--long` there is no result. I might need help with this one.

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -299,6 +315,198 @@ Conceptually, packages are overloaded.  They contain:
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+The core of an analyzer is it's ``run()`` function, which should accept no
+arguments. You can assume your analyzer has the package spec of interest at ``self.spec``
```
nice this is very clear

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -299,6 +315,198 @@ Conceptually, packages are overloaded.  They contain:
+.. code-block:: python
+
+    def save_result(self, result, outdir=None, monitor=None):
+        """Abi results are saved to individual files, so each one needs to be
```
ABI

tgamblin commented on lib/spack/spack/analyzers/__init__.py:
```diff
@@ -0,0 +1,71 @@
+]
+
+# "all" cannot be an analyzer type
+analyzer_types = {
```
look at `llnl.util.lang.list_modules()` and how we use it for most of the places where there are modular directories full o files in Spack.  e.g. `lib/spack/spack/compilers`.  Then look at `spack.util.naming.mod_to_class`. I think you can use the two of these to avoid having this table and imports that need to be updated every time there is a new analyzer.  It'll also make these consistent with a lot of the rest of spack.

vsoch replied:
This would work, I can restructure the analyzers directory to be similar to packages and compilers with one per module file.

tgamblin commented on lib/spack/spack/analyzers/abi.py:
```diff
@@ -0,0 +1,93 @@
+        Since the output for libabigail is one file per object, we communicate
+        with the monitor multiple times.
+        """
+        super().__init__(spec, dirname)
```
`super()` is python 3 -- but we need to be 2.6 compatible

tgamblin commented on lib/spack/spack/analyzers/abi.py:
```diff
@@ -0,0 +1,93 @@
+        result = {}
+
+        # Generate an output file for each binary or object
+        for obj in manifest.get("binary_to_relocate_fullpath", []):
```
nice.

tgamblin commented on lib/spack/spack/analyzers/abi.py:
```diff
@@ -0,0 +1,93 @@
+            spec = spack.spec.Spec("libabigail+docs")
+            spec.concretize()
+
+            self.abidw = spack.bootstrap.get_executable(
```
I love this.  Thanks @becker33 and @alalazo

vsoch replied:
Thank you @becker33 and @alalazo ! I already have another use case, when I can get the rpm package to build we can bootstrap debugedit.

tgamblin replied:
yeah bootstrapping `boto3` also came up on the AWS call today.

tgamblin commented on lib/spack/spack/analyzers/abi.py:
```diff
@@ -0,0 +1,93 @@
+        # We've already saved the results to file during run
+        for obj, filename in result.get(self.name, {}).items():
+
+            # Don't include the prefix
```
does this have to do with 2 lines down?  seems like yes -- can it be adjacent instead of up a line?

tgamblin commented on lib/spack/spack/analyzers/base.py:
```diff
@@ -0,0 +1,70 @@
+import os
+
+
+class AnalyzerBase:
```
`AnalyzerBase(object)` to ensure this is a new-style class in python 2 and 3

tgamblin commented on lib/spack/spack/analyzers/base.py:
```diff
@@ -0,0 +1,70 @@
+        """
+        raise NotImplementedError
+
+    def get_outfile(self, dirname=None, outfile=None):
```
can this be an `outfile` property instead of `get_outfile`?

vsoch replied:
The reason I have it as a function instead of a property (and have it named `get_outfile`) is because it accepts arguments that would determine logic for getting the output file (e.g., the directory or a different output file name). I think a property would be more appropriate if this is already set. So either we leave as a function, or we require that the user provide the exact outdir / dirname on init of the class, and then use those. Let me know your preference!

tgamblin replied:
yeah I guess this is just in the abigail analyzer, so let's keep it.  Shoudl it be a private method though?

vsoch replied:
Good thinking, I'll make it private.

vsoch replied:
Okay, so I also wasn't happy with this still - since we have consistency with an output directory (but not the file) I created a property for the output_dir, which gets generated once and then "cached" at `self._output_dir`. Then the developer user can just `os.path.join` that directory with whatever output file basename they want. I also made it so each analyzer has it's own folder so we don't have to worry about name conflicts.

tgamblin commented on lib/spack/spack/analyzers/build.py:
```diff
@@ -0,0 +1,79 @@
+        lines = re.split("(;|\n)", content)
+        lines = [x for x in lines if x not in ['', '\n', ';'] and "SPACK_" in x]
+        lines = [x.strip() for x in lines if "export " not in x]
+        lines = [x.strip() for x in lines if "export " not in x]
```
this is done twice

vsoch replied:
Copy pasta failure right there!

tgamblin commented on lib/spack/spack/analyzers/build.py:
```diff
@@ -0,0 +1,79 @@
+        lines = [x.strip() for x in lines if "export " not in x]
+        result = {}
+
+        # Dictionary comprehentions require 2.7
```
but this doesn't, and it doesn't split twice:

```python
result = dict(tuple(x.split("=")) for x in lines)
```

This assumes split only ever returns two-tuples, so I'm not sure that `split` is the right thing to use here. If the string has more than one `=`, then you probably want to use `partition()`, e.g.:

```python
result = dict(x.partition("=")[::2] for x in lines)
```

vsoch replied:
I haven't actually used partition a lot - learned something new!

tgamblin reviewed (changes requested):
Here's half a review -- I'll get through the rest tomorrow.

vsoch committed:
re-organizing analyzers to be like packages

This commit will reorganize the analyzers folder to, akin to packages,
have one analyzer per module file, with the name of the class being derivable
from the filename. This means we can add a common function, spack.util.classes,
that can take a parent path and module path and return a list of loaded modules.
This function is added, and also used to get a list of platforms in architecture.py.
This commit will also change the analyzer output to be at ~/.spack/analyzers by
default, which the user can change with config:analyzers_dir, or on the fly with
--outdir. If the user specifies an --outdir, we do not include nesting of the
package folders, but instead honor the exact choice of the user. Finally, docs
(user and developer) are updated to reflect these changes.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
forgot to capitalize ABI

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -9,30 +9,43 @@
+For all analyzers, we write to an ``analyzers`` folder in your spack ``$HOME``
+directory, or the value that you specify in your spack config at 
+``config:analyzers_dir``. We don't want to write to your package install folder
+in the case that it is read only. For example, here we see the results
```
probably don't need to include this detail about not wanting to write to the install directory.

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -9,30 +9,43 @@
-For all analyzers, we write to an ``analyze`` folder in the package install
-directory metadata folder. For example, here we see an analyze folder that is
-alongside a wget install:
+For all analyzers, we write to an ``analyzers`` folder in your spack ``$HOME``
```
> your spack ``$HOME`` directory

we usually just call this `~/.spack` -- the docs don't really introduce the concept of a "spack home directory"

trws replied:
That's fair, but probably also slightly unfortunate.  This can be set with an env var can't it? I recall the XDG issue didn't go anywhere, but is it always actually at that location?

vsoch replied:
I can change to that. I was thinking verbatim "A spack hidden directory in $HOME == spack home directory," haha.

vsoch replied:
This is fixed.

tgamblin replied:
@trws The misc cache location can be configured (not with an env var) but things outside it actually can’t.

tgamblin commented on lib/spack/docs/analyze.rst:
```diff
@@ -9,30 +9,43 @@
-    spack-configure-args.txt  install_manifest.json              spack-build-02-configure-out.txt   spack-build-env.txt
-    spec.yaml
+    $ tree ~/.spack/analyzers/
+    /home/vanessa/.spack/analyzers/
```
I guess we know you wrote this :).  In other parts of the docs -- it's a little inconsistent -- we've used `myuser`, `spackuser`, `username`, and `user`... if you want to anonymize feel free.  @alalazo is in the modules reference as `mculpo`.  Wonder if we should standardize that.

vsoch replied:
yes we probably should, if @alalazo is okay with it. I would opt for the most fun of the options (so not user or username, but maybe spackuser or similar).

vsoch replied:
I updated here to be spackuser, and we can discuss if there should be a global strategy!

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -322,73 +322,78 @@ Stage objects
-Your analyzer should be a subclass of the :class:`AnalyzerBase <spack.analyzers.AnalyzerBase>`,
-and you can look at other analyzers in that folder for examples. The guide
-here will tell you about the basic functions needed.
+To write an analyzer, you should add a new module file to the
```
I think we should say "python module" or "python file" since we have sections all about entirely different "module files"

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -322,73 +322,78 @@ Stage objects
+.. code-block:: console
+
+    $ tree ~/.spack/analyzers/
+    /home/vanessa/.spack/analyzers/
```
anonymize?

vsoch committed:
making get_outdir for an analyzer private -> _get_outdir

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

tgamblin commented on lib/spack/spack/analyzers/libabigail.py:
```diff
@@ -0,0 +1,93 @@
+        """
+        super(Libabigail, self).__init__(spec, dirname)
+
+        # This doesn't seem to work to import on the module level
```
For most of this stuff, I think we do not want to import at module level, since if it is an optional dependency it should only be bootstrapped *if* we need the functionality.  If you import at module level, it'll always be bootstrapped if you import the analyzer, and I think we want to be lazier than that.

vsoch replied:
So this implementation is correct?

tgamblin replied:
LGTM!

tgamblin commented on lib/spack/spack/architecture.py:
```diff
@@ -496,23 +495,8 @@ def arch_for_spec(arch_spec):
-        classes.append(cls)
-
-    return classes
+    return spack.util.classes.list_classes("spack.platforms", mod_path)
```
making the world a slightly better place :)

tgamblin commented on lib/spack/spack/cmd/analyze.py:
```diff
@@ -0,0 +1,110 @@
+    subparser.add_argument(
+        '-o', '--outdir', default=None,
+        dest='outdir',
+        help="write output to a different directory than ~/.spack/analyzers")
```
`spack stage` has `-p` / `--path` -- should the UI be consistent with that?

vsoch replied:
Yes I think so.

tgamblin commented on lib/spack/spack/cmd/analyze.py:
```diff
@@ -0,0 +1,110 @@
+    subparser.add_argument(
+        '--list-analyzers', action="store_true",
+        default=False,
+        help="list available analyzers.")
```
list and analyze are really different modes here, and everywhere else in spack we use subcommands for this kind fo thing.

I think you want `spack analyze list-analyzers` and `spack analyze run [all|analyzer [...]]`. Also might be good to have `spack analyze list` to show available results (like `spack test`), though that can probably come later.

With subcommands, the options then become less confusing, as `--overwrite` becomes an option on `run` and not on `list-analyzers`, so you can't get weird combinations of args.

vsoch replied:
Good idea - I'll add these as subcommands proper. Let's hold off on adding `list` until we have a good understanding of what we want to show. But I do think I should re-organize analyzer output into it's own namespace, so each analyzer can output whatever content it wants to that directory without worrying about over-writing another analyzer's files.

vsoch replied:
I also think the user might want to run more than one spec, or more than one analyzer, so it makes sense to have as many specs listed as needed, and still take analyzer names with `-a`. E.g.,

```bash
# Run libabigail and diff analyzers on axom and ascent
$ spack analyze run -a libabigail -a diff axom ascent
```

tgamblin replied:
Oh I like that a lot

tgamblin commented on lib/spack/spack/environment.py:
```diff
@@ -445,23 +445,45 @@ def _write_yaml(data, str_or_file):
 
-def _eval_conditional(string):
-    """Evaluate conditional definitions using restricted variable scope."""
+def get_host_environment_metadata():
```
it's a little weird for this to be in `environment.py` (which is about spack environments), and not in `spack.architecture` or something like that.  It looks like _get_host_environment was already here but maybe this stuff goes somewhere else.

vsoch replied:
Yeah that's why I added it here, because it's alongside `_get_host_environment`. You want to move both?

tgamblin replied:
yeah I think both go either in `spack.util.environment` or in `spack.architecture`

vsoch replied:
Moved to spack.util.environment.

tgamblin commented on lib/spack/spack/environment.py:
```diff
@@ -1486,6 +1510,11 @@ def install_all(self, args=None, **install_args):
             builder = PackageInstaller(installs)
+
+            # If the user has requested a monitor client, attach to builder
+            if install_args.get('use_monitor', False):
```
I wish we had a single place to put this and other preinstall stuff.  But maybe this is fine for now.

vsoch replied:
Yeah it's probably not ideal to be passing around install-args! What did you have in mind?

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -572,6 +579,21 @@ def log(pkg):
     dump_packages(pkg.spec, packages_dir)
 
 
+def combine_phase_logs(pkg):
```
I'm actually ok with not doing this and just shipping separate logs.  We could just make this change for 0.17, though we'd need to modify some of the build and error output to point to the per-phase logs.

@cyrush @white238: are you going to be ok with separate build phase logs?  Or is that going to break uberenv an would you rather have a combined log as well?  This PR allows the logs to be split by phase, and for us to do things like see configure args and other build metadata independently.

If we add nice phase timings for @cyrush does that make you even cooler with this?

vsoch replied:
I added this because I was worried that if a user might depend on that file, removing it would be an issue. I don't think it hurts to keep it.

white238 replied:
I prefer separate build phase logs.

cyrush replied:
separate build phase logs actually sounds good to me.  
Addition of timestamps would be be awesome as well.

tgamblin replied:
@vsoch: let's get rid of the old log given that @cyrush is ok w/this.  Some of the logs can be quite large, so I think it makes sense not to store them twice.

vsoch replied:
Gotcha - will remove! I'm writing the hook functionality now (and also lots of docs) so I will get this in soon (but probably not immediately)

vsoch replied:
@tgamblin removing this is adding a bunch of funky errors - I think it should be done separately from this PR, it's out of scope and can be easily done after.

vsoch replied:
Just to update everyone here - I'm going to leave in the "global log" for now because removing it requires updating several other tests and modules that expect it, and I think the scope of this PR is getting too large (but it will be easy to do, and more cleanly, after in a separate one).

trws replied:
Perhaps create an issue referencing this to make sure that doesn't get dropped on the floor?

vsoch replied:
All set: https://github.com/spack/spack/issues/22399 And I'll leave the discussion here un-resolved so if anyone clicks the link to it, the discussion too is not lost in the doldrums of a resolved comment thread.

tgamblin replied:
thanks!

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -1550,6 +1618,7 @@ def install(self):
                 # Only terminate at this point if a single build request was
                 # made.
                 if task.explicit and single_explicit_spec:
+                    self._send_task_failure(task.request.pkg.spec)
```
I'm worried in general about the shotgun mods to this file. I think there are too many concerns in installer.

Is it possible to raise a general type of exception for failures (i.e. wrap the exception if it's not a type you can catch all in one `except` block) and send the task failure someplace *outside* the core logic? I just think there is too much going on here and it'll become unmaintainable.

vsoch replied:
This might be out of scope for the PR here - I just looked at this section again and it's covered in try/except/finally so I think we would need to rethink the entire structure. The only changes I added don't modify this structure (even if it's not great), they just insert the monitor to send an error in (the many!) places that can occur. I'm not familiar enough with this section to know how to best delete all these try/excepts in favor of one massive one. But if someone else familiar wants to walk it through with me, I could help! But a major refactor of this logic feels out of scope for this PR, which is already too big :)

tgamblin replied:
Yeah let's not rework the exception structure -- but see below w.r.t hooks.

vsoch replied:
I created an issue to track this (so we don't forget) https://github.com/spack/spack/issues/22400

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -1641,6 +1714,16 @@ def build_process(pkg, kwargs):
+    monitor = None
+
+    # Instantiate a monitor for the worker, if requested
+    if use_monitor:
```
similarly curious here whether monitors could be implemented via hooks -- can we have a general set of hooks here and not an `if monitor` for this feature (and the next feature and the next one ...)

vsoch replied:
I'm not familiar with how hooks work, is there an example somewhere?

vsoch replied:
## Did somebody say they wanted a hook?
![image](https://user-images.githubusercontent.com/814322/111557156-c1e45700-8751-11eb-8508-982c62caa6dd.png)

tgamblin replied:
Look here: https://github.com/spack/spack/blob/develop/lib/spack/spack/hooks/__init__.py and at all the modules in that directory.

Basically, instead of sticking monitor-specific calls here, I think `on_install_failure(exception)` or something like that should be a new hook for these spots in the code.  Then we can write other things that need to handle the same events.

A hook is just a callback.  It's yet another way to do the observer pattern.

tgamblin replied:
Also the idea is to get all the monitor code in one place -- right now it's scattered all over installer.

vsoch replied:
Okay so two questions:

1. If the hook just accepts an exception, how do we also give it other variables like monitor preferences from the user?
2. Given something like the monitor, how do we instantiate it once (and then use over again) so we don't have to do it every time there is a hook (related to 1.)

tgamblin replied:
1. Not tied to the arguments -- that's up to you.  I think making it take a spec as well makes a lot of sense.
2. In the hook module, I think you should create an object and store it when installation starts (so the first hook would create the object).  Just spitballing here, but I might store in a `dict` in the hook module, keyed by spec.

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -1715,9 +1810,24 @@ def build_process(pkg, kwargs):
+                        try:
+                            phase(pkg.spec, pkg.prefix)
+                            if monitor:
+                                monitor.send_phase(pkg, phase_name, log_file, "SUCCESS")
```
this and the sends of failures above could be callouts to some type of general hook instead of just something for monitor -- is that easy to do ?

vsoch replied:
If I knew what a hook was, maybe? :P

tgamblin replied:
see above

tgamblin reviewed (changes requested):
ok working my way through even more.  More comments -- some of the core stuff is getting a bit complicated so I wonder if we can reduce the cognitive load for maintainers in `installer.py`

trws commented on lib/spack/docs/monitoring.rst:
```diff
@@ -0,0 +1,88 @@
+
+.. code-block:: console
+
+    ==> Error: You are required to export SPACKMON_TOKEN and SPACKMON_USER
+
+This extra check is to ensure that we don't start any builds,
+and then discover that you forgot to export your token. However, if 
+your monitoring server has authentication disabled, you can tell this to
+the client to skip this step:
```
Does this step check if the token and user are valid, so the messages are combined, or does it check for existence of the env vars without poking the interface? I'm not sure I'd ask for this to change right now either way, but if possible it would be nice for an auth-disabled endpoint to be discovered so it just works.

vsoch replied:
It checks for the environment variables, and pings the monitor endpoint to check that the server is running. Since the server will ask for authentication when it's warranted (e.g., the information endpoint doesn't need auth, and the server can choose to customize which endpoints have it) we wouldn't exit with an authentication failure message until the first ping to the server, which is right after the monitor instance is created when we try to create a new Build, [here](https://github.com/spack/spack/pull/21930/files#diff-5c111ba1f5bfb321d12a190c2a79b33c7b53e3d02f6d0201f1703432b71cb17cR406). There is no global "auth disabled endpoint" other than the service info endpoint at `/m1/` that we ping at the beginning, so there isn't a work around for if auth fails unless the server doesn't require it at all. Given that auth is required for any endpoint except for service info, it will exit on error if the authentication fails, which is what I think we'd want if the user provides credentials that they expect to work. If the user wants to say "Try the endpoint and I don't care if it works or not, keep going) then they can use the flag `--monitor-keep-going` and it will just [keep going](https://github.com/spack/spack/pull/21930/files#diff-7f0454509928f4f8fee4c25eba560f8bc4f7a5fbc6aeef8a31075041a6df4133R215). The user should generally know if their spack monitor server requires auth, and if it does, be able to get tokens for it.

trws replied:
Ok, works for me, thanks!

trws commented on lib/spack/spack/analyzers/configargs.py:
```diff
@@ -0,0 +1,27 @@
+class Configargs(Analyzerbase):
+
+    name = "config_args"
+    outfile = "spack-analyzer-install-files.json"
```
Is this where you want these? Just noticing it's the same as the installfiles path.

vsoch replied:
oups, that's a typo :P

trws commented on lib/spack/spack/analyzers/environmentvariables.py:
```diff
@@ -0,0 +1,48 @@
+    description = "environment variables parsed from spack-build-env.txt"
+
+    def run(self):
+        """Given a directory name, return the json file to save the result to
+        """
```
```suggestion
        """Given a spec with an installed package, run the analyzer on it.
         """
```

vsoch replied:
Actually if it's okay with you I'll write something much more descriptive than this (I didn't update them after separating them and the descriptions are awful!)

trws replied:
Sounds great to me. Just saw what looked like a swap.

trws commented on lib/spack/spack/analyzers/libabigail.py:
```diff
@@ -0,0 +1,93 @@
+
+    def run(self):
+        """Run libabigail, and save results to filename. This run function differs
+        in that we write as we generate and then return a lookup of files.
```
Lookup of files?

vsoch replied:
This is poorly described! I'll fix this too.

trws commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -551,52 +551,69 @@ def read_buildinfo_file(prefix):
             if relocate.needs_text_relocation(m_type, m_subtype):
-                rel_path_name = os.path.relpath(path_name, prefix)
-                text_to_relocate.append(rel_path_name)
+                data['text_to_relocate'].append(rel_path_name)
```
Does this not count as `added`?

vsoch replied:
You're right, it should be.

vsoch replied:
Okay, added is added :laughing:

vsoch committed:
updating spack analyze organization and base

this commit will update analyers to output to a scoped directory (named after the
analyzer) within ~/.spack/analyzers, instead of outputting to the same folder. We
are doing this to avoid different analyzers writing the same named files that overwrite one
another. We are also adding the --overwrite argument, along with subcommands for run and
list-analyzers to better separate the command namespace. The docs should be updated, along
with a vermin issue in analyzers/__init__.py. We are also updating --outdir to be --path so
that the analyzer module is consistent with spack stage. I want (and need) to write a test
module for analyze, which should be reasonable to do since we save to the filesystem. I
did not do it previously because I expected needing interaction with a monitor server,
which of course is not required.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
missing a run with spack analyze

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

trws commented on lib/spack/spack/test/cmd/install.py:
```diff
@@ -142,7 +142,8 @@ def test_install_output_on_build_error(mock_packages, mock_archive, mock_fetch,
     # capfd interferes with Spack's capturing
     with capfd.disabled():
-        out = install('build-error', fail_on_error=False)
+        out = install('-v', 'build-error', fail_on_error=False)
+
```
Did you mean to leave this in?

vsoch replied:
Yes, I think @tgamblin mentioned in slack that if we add the `-v` option the test works (but I haven't seen it run yet).

trws commented on lib/spack/spack/spec.py:
```diff
@@ -1642,7 +1642,12 @@ def to_node_dict(self, hash=ht.dag_hash):
+            package_hash = self.package.content_hash()
+
+            # Full hashes are in bytes
+            if isinstance(package_hash, bytes):
```
Is this intended to convert all hashes to `unicode` type in python 2.7?  It may not actually affect anything, but I'm always wary of `isinstance(..., bytes)` after some issues with 2/3 compat in flux.

vsoch replied:
Is there a better way to do it? The full hash (unlike the others) is in bytes so we do need to convert it (somehow!)

trws replied:
The trick is that `bytes` and `str` are the same type in python2.7, so this will convert all strings in python 2 to the `unicode` type but only `bytes` instances to `str` in python3.  As long as it's ok that all of them come out unicode I'd probably leave it like this, perhaps with a comment about it.  If we want everything to be the native text type `str` regardless of the python version, then the check should probably be something like `if (not isinstance(package_hash, six.text_type)) and isinstance(package_hash, six.binary_type)`.

vsoch replied:
I'm happy to change that! @trws do you have familiarity with hooks and know how we could keep state, for something like spack monitor?  My questions here -> https://github.com/spack/spack/pull/21930#discussion_r597056059. It's what I'm working on now and need to figure out before updating the PR.

trws reviewed (commented):
This looks like great stuff overall to me.  A few questions in here, some typos, but mostly good stuff.

vsoch committed:
updating with review changes from @trws

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
@trws @tgamblin I'll write an analyzer test module tomorrow, I actually have one started locally and erm, sort of forgot about it, heh.

vsoch committed:
adding monitor interactions via hooks

This commit includes re-writing monitor calls in installer.py to instead
use newly added hooks, including:

on_install_failure
on_install_success
on_phase_success
on_phase_error
on_analyzer_save
on_install_start

These functions are used in the new hooks for the monitor at lib/spack/spack/hooks/monitor
and I have also added a documentation section for the developer that explains each
hook type and how to write new ones.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
Okay for anyone interested, hooks are added! I tested locally with install and analyze, both with and without monitor, but still need to write tests proper. https://github.com/spack/spack/pull/21930/commits/636a8fa6eee066f84d6872d6a4bd54e92c70a271

trws commented:
Hook setup looks nice!  I think all that leaves is hopefully hoisting the calls of the `on_install_failure` to a common spot if possible.

vsoch commented:
@trws agreed, but that requires a refactor of installer.py which is outside the scope of this PR.

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -563,6 +575,182 @@ Whenever you add/remove/rename a command or flags for an existing command,
+^^^^^^^^^^^^^^
+
+The following hooks are currently implemented. If there is a hook that you
+would like and is missing, you can propose to add a new one.
```
how? maybe just say this makes it easy to write hooks?

vsoch replied:
Added.

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -563,6 +575,182 @@ Whenever you add/remove/rename a command or flags for an existing command,
+    post_log_write = HookRunner('post_log_write') # <- here is my new hook! 
+    
+
+I would then need to decide what arguments my hook would expect. Since this is
```
this may seem nitpicky but most of the docs are 2nd person -- can we stick to that here?  The other sections in here seem to use "you" as well.

vsoch replied:
Why are they in second person?

vsoch replied:
Oh wait nevermind, it's because I used "I" - gotcha.

vsoch replied:
Not sure why I switched into I statements, I agree you is better!

vsoch replied:
Fixed.

trws commented:
If it's a material refactor, definitely no need to pull it in here. :shipit:
Thanks @vsoch.

tgamblin reviewed (changes requested):
smaller requests.

vsoch committed:
adding basic tests for analyzers

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
docs should be in second person (no Is)

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
init monitor function no longer exists on installer

and I am also removing some added new lines that we definitely do
not need

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
developer guide heading title too short

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
updating test for install on error

the previous test (when all output was directed to one log)
would return the output for th entire set of phases, but since
we have updated spack to output to different logs based on
phases, this test now only gets the last phase. I am updating
the test to reflect this.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
test no longer has configure error in output

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
@trws and @tgamblin the failing test now looks like the same method is used to derive output for cdash - I'm not familiar with how this works and what is expected so i absolutely don't want to mess up changing it:

```bash
2.23s call     lib/spack/spack/test/cmd/env.py::test_env_updates_view_force_remove[mock_archive0]
2.18s call     lib/spack/spack/test/cmd/deprecate.py::test_deprecate_deps[mock_archive0]
=================================== FAILURES ===================================
_________________ test_cdash_upload_build_error[mock_archive0] _________________

tmpdir = local('/tmp/pytest-of-spack-test/pytest-0/test_cdash_upload_build_error_0')
mock_fetch = None, install_mockery = None
capfd = <_pytest.capture.CaptureFixture instance at 0x8f41248>

    @pytest.mark.disable_clean_stage_check
    def test_cdash_upload_build_error(tmpdir, mock_fetch, install_mockery,
                                      capfd):
        # capfd interferes with Spack's capturing
        with capfd.disabled():
            with tmpdir.as_cwd():
                with pytest.raises((HTTPError, URLError)):
                    install(
                        '--log-format=cdash',
                        '--log-file=cdash_reports',
                        '--cdash-upload-url=http://localhost/fakeurl/submit.php?project=Spack',
                        'build-error')
                report_dir = tmpdir.join('cdash_reports')
                assert report_dir in tmpdir.listdir()
                report_file = report_dir.join('Build.xml')
>               assert report_file in report_dir.listdir()
E               AssertionError: assert local('/tmp/pytest-of-spack-test/pytest-0/test_cdash_upload_build_error_0/cdash_reports/Build.xml') in [local('/tmp/pytest-of-spack-test/pytest-0/test_cdash_upload_build_error_0/cdash_reports/Update.xml')]
E                +  where [local('/tmp/pytest-of-spack-test/pytest-0/test_cdash_upload_build_error_0/cdash_reports/Update.xml')] = <bound method LocalPath.listdir of local('/tmp/pytest-of-spack-test/pytest-0/test_cdash_upload_build_error_0/cdash_reports')>()
E                +    where <bound method LocalPath.listdir of local('/tmp/pytest-of-spack-test/pytest-0/test_cdash_upload_build_error_0/cdash_reports')> = local('/tmp/pytest-of-spack-test/pytest-0/test_cdash_upload_build_error_0/cdash_reports').listdir
```
Do we need to re-think how the same (full output) would be delivered to cdash? Or is this part of the task too complicated for this first PR, and we should revert back to having output be in one log, and I'll update spack monitor to not represent phases? Let me know what you'd like to do!

tgamblin commented:
> Do we need to re-think how the same (full output) would be delivered to cdash? Or is this part of the task too complicated for this first PR, and we should revert back to having output be in one log, and I'll update spack monitor to not represent phases? Let me know what you'd like to do!

The CDash and JUnit reporters just read `spack-build.out` immediately after `_install_task`.  I tried this locally and it looks like there is a race somewhere (or maybe a failure to flush?) in how your logging is done.  It seems to fail most of the time but succeeds sometimes.

This seems like a symptom, not a cause, so do you have unit tests for how you are combining the logs?  It doesn't look like there's anything to verify that the function that combines the logs works.

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -412,6 +413,20 @@ def clear_failures():
+    log_files = pkg.phase_log_files
+
+    with open(pkg.log_path, 'w') as fd:
+        for line in itertools.chain.from_iterable(list(map(open, log_files))):
```
are these ever closed?

vsoch replied:
Totally missed this. No I don't think so.

tgamblin commented on lib/spack/spack/installer.py:
```diff
@@ -412,6 +413,20 @@ def clear_failures():
+    """
+    log_files = pkg.phase_log_files
+
+    with open(pkg.log_path, 'w') as fd:
```
`fd` is a bit of a weird variable name here as it's typically used for file descriptors (low-level ints returned by `os.open()`).  I would use `log_file` or something here, as this is really a file object.

vsoch replied:
Will do. I always uesd "filey" when I was new to Python, but then always say "fd" for file descriptor, and fh for file handle, and thought that was best practice.

vsoch commented:
@tgamblin the only change to logging I did was to reverse the loops so the logging is on the inner loop and takes a phase log file instead of having one master log file on the outside writing to one file. I don't know what's wrong so I'm going to undo that change and just refactor spack monitor to accept one final log. Someone else that understands how logging works can take a closer look at this, I don't think I can fix it.

vsoch committed:
removing reversal of output logs

there is some race condition that is causing errors left and right,
not just for install, but just about everything that uses it (e.g.,
reporting like cdash). I added a test for combining phase logs and also
refactored the function for the files to be properly opened and closed
with "with," but the error persists. This update was a "nice to have"
feature of the PR, but is not actually essential for it, and I think
it is too big of bite to take trying to debug the complicated logging
when it is really not related to the analyzers or spack monitor,
which both can work fine without the logs being separate. Thus I am
removing this change from this PR because it needs to be addressed
separately, and likely by someone that understands the complexity of
the llnl log util.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

tgamblin commented:
@v: we really want the split logs for this so I’d rather have that. Also I think we should dig in and figure out that issue. I can help if you want.

vsoch commented:
@tgamblin I propose getting this working and merged so I'm not held back for working on the diff analyzer (requires this analyzer base to work), and then addressing splitting the logs separately. We could also then talk about and start to work on having cdash / reporters run using hooks instead of the current setup. I don't think splitting the logs is absolutely essential for this PR, but that doesn't mean we shouldn't do it, and soon. Is there any reason that wouldn't work?

vsoch committed:
Path type needs to be string to use os.path.join (not sure why it worked locally when I wrote the test

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

tgamblin committed:
revert "removing reversal of output logs"

This reverts commit a86e4d7269df7480162a5c683b4529beab467425.

tgamblin committed:
bugfix: fix race between combining phase logs and log_output

Anything that needs to process logs should happen *after* `log_output`
completes, so we need the call to `combine_phase_logs` to be outside the
`with` block in `_install_task`.

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -299,6 +325,210 @@ Conceptually, packages are overloaded.  They contain:
+An analyzer run Function
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+The core of an analyzer is it's ``run()`` function, which should accept no
```
its

tgamblin commented on lib/spack/spack/test/cmd/analyze.py:
```diff
@@ -0,0 +1,125 @@
+
+def test_environment_analyzer(tmpdir, mock_fetch, install_mockery_mutable_config):
+
+    """test the environment variables analyzer
```
are these docstrings intentionally one line below the function name?  Or shoud they be just under it?

vsoch replied:
I think I probably started with two lines of args, like:
```python
def test_install_output_on_build_error(mock_packages, mock_archive, mock_fetch,
                                       config, install_mockery, capfd):
    """This test used to assume receiving full output, but since we've updated
```
and then removed most of the pytest fixures and left the line empty, and then proceeded to copy paste it. Will fix asap!

tgamblin commented on lib/spack/docs/developer_guide.rst:
```diff
@@ -299,6 +325,210 @@ Conceptually, packages are overloaded.  They contain:
+                ├── install_files
+                │   └── spack-analyzer-install-files.json
+                └── libabigail
+                    └── spack-analyzer-libabigail-libz.so.1.2.11.xml
```
this probably doesn't need to be fixed in this PR but I think we'll want to mirror the original package's directory structure under libabigail to ensure each library gets a unique path.  I'm guessing some package will come up with duplicate lib and/or executable names.

vsoch replied:
This should be fixed, and I added a note in the developer docs that if they have object-specific files, they should honor the structure.

tgamblin commented on lib/spack/spack/analyzers/analyzerbase.py:
```diff
@@ -0,0 +1,105 @@
+
+
+def get_analyzer_dir(spec, analyzer_dir=None):
+    """Given a spec, return the directory to save analyzer results,
```
We're not super religious about docstrings but FYI, if you want the docs to look nice, a one-line summary with a period followed by the rest of the text is what PEP257 says you're supposed to do: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings.   I think some doc formatters care.

vsoch replied:
I'd like to do them right. I'll go through the PR (probably tomorrow, need to have dinner) and fix them up.

vsoch replied:
Another question - what standard do you use for commits? I notice you do:

```
<category>: short message

longer message
```
And I'd like to do that too - I currently do something similar but without the category. I've seen you use "bugfix" as a category, but how do I know what the others are? (I'd guess there is docs too).

vsoch committed:
ensuring that phase log files are closed when being read, and combine_phase_log function works with two arguments

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

tgamblin commented on lib/spack/spack/analyzers/analyzerbase.py:
```diff
@@ -0,0 +1,105 @@
+    def output_dir(self):
+        """The full path to the output directory, including the nested analyzer
+        directory structure. This will create the full path if it doesn't exist,
+        so you should be sure that you have a result before using the property.
```
the auto-creation thing is a little weird -- I'd expect this to return the path regardless of whether it exists, and save_result to create the directory.  I think it'll be easy to create superfluous directories with the functionality merged like this. Is there a reason you want the property to create the directory?

vsoch replied:
You can probably tell from my justification in the docstring that I thought about it, and did it this way (but wasn't very happy with the prospect of having directories created) and then added "you should have a result before using this property." I think you are probably right we should have the save_result function do the creation, I'll fix this up.

vsoch replied:
This should also be fixed. The save_result (or wherever the analyzer needs to write files) is in charge of creation.

tgamblin commented on lib/spack/spack/analyzers/environmentvariables.py:
```diff
@@ -0,0 +1,50 @@
+import re
+
+
+class Environmentvariables(Analyzerbase):
```
any reason not to call this `environment_variables.py` and the class `EnvironmentVariables`?

vsoch replied:
I like that better.

vsoch replied:
I'm going to do it for the others too.

tgamblin commented on lib/spack/spack/analyzers/environmentvariables.py:
```diff
@@ -0,0 +1,50 @@
+        lines = [x.strip() for x in lines if "export " not in x]
+
+        # Dictionary comprehentions require 2.7
+        return dict(x.partition("=")[::2] for x in lines)
```
I thought about this some more and I think this will actually fail for many environments -- particularly ones with functions in them.  Function decls in bash show up as multi-line things in the source-able environment file.

You could just source the file:

```
from spack.util.environment import EnvironmentModifications
mods = EnvironmentModifications.from_sourcing_file(filename)
env = {}
mods.apply(env)
return env
```

The only thing I am not sure of is whether you'll miss part of the env this way (because some variables are unchanged from spack's environment).  It may be that we want to have a json dump in addition to the sourceable 
one to avoid all this -- as it's ultimately spack that dumps the file this is reading, anyway.

vsoch replied:
okay just testing this out - it works well but adds a bunch of errors. It's probably okay.

trws replied:
Does spack export functions into these files? That's a behavior I haven't seen.

tgamblin commented on lib/spack/spack/cmd/install.py:
```diff
@@ -106,6 +107,8 @@ def setup_parser(subparser):
         '--cache-only', action='store_true', dest='cache_only', default=False,
         help="only install package from binary mirrors")
 
+    monitor_group = spack.monitor.get_monitor_group(subparser)  # noqa
```
does this need noqa?

vsoch replied:
Yes, if I don't have it, flake8 tells me there is an unused variable.

vsoch replied:
I can take out the variable and add a comment, that should work and not need the noqa. I left it there because I want the developer to know what the heck it does.

tgamblin reviewed (changes requested):
few last things

vsoch committed:
updating organization and naming of spack analyze modules

This commit will also update the docstrings to have one line with a sentence,
followed by a newline, then remaining lines after that. We also are using
the EnvironmentModifications class to parse an environment file, and ensuring
that libabigail object output is written to the object respective directories

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
forgot to add in renamed files!

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
The tests ran immediately, awesome! I need to do more work in off hours and weekends :smile:

tgamblin commented:
You don’t have to work on weekends — you can just thank @alalazo for #22410, #22421, #22426, and #22430, which tamed the queue and got us back to productive CI. 😃

vsoch commented:
Thank you @alalazo !

vsoch committed:
adding more tests for analyzer, and moving get_host_environment and get_host_environment_metadata into spack.util.environment

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
remove unused import

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
see if we can mock install python@3.8 for libabigail

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
libabigail analyzer test is not going to work

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
monitor group needs to be regular (not mutually exclusive) group

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
resolving merge conflict

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
adding debugging info for lock

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
remove extra prints, the error did not trigger again

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch committed:
try using socket.gethostname over socket.getfcdn

Signed-off-by: vsoch <vsoch@users.noreply.github.com>
